mdsn/src/lib.rs

```
@@ -263,111 +269,289 @@ pub struct Dsn {
         };
-        let username = cap
-            .name("username")
-            .map(|m| {
-                let s = m.as_str();
-                urlencoding::decode(s)
-                    .map(|s| s.to_string())
-                    .map_err(|err| DsnError::InvalidPassword(s.to_string(), err))
-            })
-            .transpose()?;
-        let password = cap
-            .name("password")
-            .map(|m| {
-                let s = m.as_str();
-                urlencoding::decode(s)
-                    .map(|s| s.to_string())
-                    .map_err(|err| DsnError::InvalidPassword(s.to_string(), err))
-            })
-            .transpose()?;
-        let subject = cap.name("subject").map(|m| m.as_str().to_string());
-        let path = cap.name("path").map(|m| m.as_str().to_string());
-        let addresses = if let Some(addr) = cap.name("addr") {
-            let addr = addr.as_str();
-            if addr.is_empty() {
-                vec![]
-            } else {
-                let mut addrs = Vec::new();
 
-                for s in addr.split(',') {
-                    if s.is_empty() {
-                        continue;
+        #[inline]
+        fn percent_decode(encoded: String) -> String {
+            decode(&encoded).map(|s| s.to_string()).unwrap_or(encoded)
+        }
+        type ParsedAddr = (Option<String>, Option<String>, Option<String>, Vec<Address>);
+        fn parse_addr(main: &str) -> Result<ParsedAddr, DsnError> {
+            if main.trim().is_empty() {
+                return Ok((None, None, None, vec![]));
+            }
+            let mut protocol = None;
+            if let Some(at) = main
+                .bytes()
+                .enumerate()
+                .rev()
+                .find(|(_, v)| *v == b'@')
+                .map(|(at, _)| at)
+            {
+                let (user_pass, addrs) = main.split_at(at);
+                let mut addrs = addrs.strip_prefix('@').expect("strip prefix @");
```

Wouldn't it be better to return an error instead of expect?

---

use `split_at` by '@' position, so `expect` may be better than a never reachable error.

---

You are correct, I overlooked the `find` above.